primitives/src/currency.rs

```diff
@@ -0,0 +1,193 @@
+use codec::{Decode, Encode};
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+use sp_runtime::RuntimeDebug;
+use sp_std::convert::TryFrom;
+use sp_std::{prelude::Vec, vec};
+#[allow(unused_imports)]
+use CurrencyType::*;
+
+macro_rules! create_currency_id {
+	($(#[$meta:meta])*
+	$vis:vis enum TokenSymbol {
+        $($(#[$vmeta:meta])* $symbol:ident($name:expr, $deci:literal, $ctype:ident) = $val:literal,)*
+    }) => {
+        $(#[$meta])*
+        $vis enum TokenSymbol {
+            $($(#[$vmeta])* $symbol = $val,)*
+        }
+
+		impl TryFrom<u8> for TokenSymbol {
+            type Error = ();
+
+            fn try_from(v: u8) -> Result<Self, Self::Error> {
+                match v {
+                    $($val => Ok(TokenSymbol::$symbol),)*
+                    _ => Err(()),
+                }
+            }
+        }
+
+		impl GetDecimals for CurrencyId {
+			fn decimals(&self) -> u32 {
+				match self {
+					$(CurrencyId::Native(TokenSymbol::$symbol) => $deci,)*
+					$(CurrencyId::UnderlyingAsset(TokenSymbol::$symbol) => $deci,)*
+					$(CurrencyId::WrappedToken(TokenSymbol::$symbol) => $deci,)*
+				}
+			}
+		}
+
+		impl CurrencyId {
+			pub fn get_enabled_tokens_in_protocol(token_type: CurrencyType) -> Vec<CurrencyId> {
+				let mut enabled_tokens = vec![];
+				$(
+					if token_type == $ctype {
+						enabled_tokens.push(
+							match $ctype {
+								Native => CurrencyId::Native(TokenSymbol::$symbol),
+								UnderlyingAsset => CurrencyId::UnderlyingAsset(TokenSymbol::$symbol),
+								WrappedToken => CurrencyId::WrappedToken(TokenSymbol::$symbol),
+							}
+						);
+					}
+				)*
+				enabled_tokens
+			}
+		}
+
+		$(pub const $symbol: CurrencyId = match $ctype {
+			Native => CurrencyId::Native(TokenSymbol::$symbol),
+			UnderlyingAsset => CurrencyId::UnderlyingAsset(TokenSymbol::$symbol),
+			WrappedToken => CurrencyId::WrappedToken(TokenSymbol::$symbol),
```

Comment:
This one operates with the so called ticker - the name of asset. Plus we have tokens and asset here, and it is not clear that they are related. I'd suggest:
```suggestion
                        UnderlyingAssetTicker => CurrencyId::UnderlyingAssetTicker(TokenTicker::$symbol),
			WrappedAssetTicker => CurrencyId::WrappedAssetTicker(TokenTicker::$symbol),
```

Reply:
https://ru.wikipedia.org/wiki/%D0%A2%D0%B8%D0%BA%D0%B5%D1%80